File: nodejs/src/Tree/MaxDepth/MaxDepth.ts
import BaseTreeAlgorithmT from "../BaseTreeAlgorithm";
import TreeNodeT from "../TreeNode";

/**
 * A binary tree's maximum depth is the number of nodes along the longest path
 * from the root node down to the farthest leaf node.
 */
class MaxDepth<T> extends BaseTreeAlgorithmT<T>{

  exec(): number {

    const depth = (node: TreeNodeT<T> | null): number => {
      // base case
      if(node === null) return 0;

      // current depth is left depth or right depth + 1
      return Math.max(depth(node.left), depth(node.right)) + 1
    }

    return depth(this.tree.root)
  }
}

export default MaxDepth;

File: go/src/tree/max_depth/max_depth.go
package max_depth

import (
	"algorithms/go/tree"
)

/*
*
The max depth of a node is the max depth of left & right + 1
*/
func maxDepth(root *tree.Node) int {
	if root == nil {
		return 0
	}
	return max(maxDepth(root.Left), maxDepth(root.Right)) + 1
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}
